engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+            } else if (node.getValue().getKey().equalsIgnoreCase("layoutInfo")) {
+                populateLayoutContextMenu(node, addLevel);
+            } else {
+                populateGenericContextMenu(node, addLevel);
+            }
+        }
+    }
+
+    private void populateWidgetContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        // Get the ClassMetadata of the widget with the given type.
+        String type = (String) node.getChildWithKey("type").getValue().getValue();
+        ClassMetadata<? extends UIWidget, ?> elementMetadata = nuiManager
+            .getWidgetMetadataLibrary()
+            .resolve(type, ModuleContext.getContext());
+
+        if (elementMetadata != null) {
+            for (FieldMetadata fieldMetadata : elementMetadata.getFields()) {
+                Field field = fieldMetadata.getField();
+                field.setAccessible(true);
+
+                // Add options for all @LayoutConfig-annotated fields.
+                if (field.isAnnotationPresent(LayoutConfig.class)) {
+                    String name = getNodeName(field);
+                    if (!node.hasChildWithKey(name)) {
+                        addLevel.addOption(name, n -> {
+                            try {
+                                createChild(name, node, field, elementMetadata);
+                            } catch (IllegalAccessException e) {
+                                return;
+                            }
+                            addContextMenuListeners.forEach(UpdateListener::onAction);
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown widget type: {}", type);
+        }
+    }
+
+    private void populateLayoutContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        String type = "";
+        JsonTree currentNode = node;
+        ClassMetadata layoutMetadata = null;
+        Class layoutHintType = null;
+
+        // Iterate down the tree and attempt to find a UILayout node.
+        while (layoutHintType == null) {
+            if (currentNode.hasSiblingWithKey("type")) {
+                type = (String) currentNode.getSiblingWithKey("type").getValue().getValue();
+                layoutMetadata = nuiManager.getWidgetMetadataLibrary().resolve(type, ModuleContext.getContext());
+
+                // If one is found, attempt to get the layout hint type.
+                try {
+                    layoutHintType = (Class) ReflectionUtil
+                        .getTypeParameter(layoutMetadata.getType().getGenericSuperclass(), 0);
+                } catch (NullPointerException e) {
+                    layoutHintType = null;
+                }
+            }
+            if (layoutHintType == null) {
+                if (currentNode.isRoot()) {
+                    break;
+                }
+                currentNode = (JsonTree) currentNode.getParent();
```

Comment:
Agreed, but IMO the method doesn't need any changes since the recursive search will always stop at the second child anyway.
In the future sufficiently complex widgets (i.e. with widget fields of widget fields) might actually require recursion.